File: geco_irig_decode.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define SAMPLE_RATE 16384           /* sample rate of ADC */
#define BITS_PER_SECOND 100         /* bits per second in IRIG-B spec */
#define HIGH_SIGNAL_THRESHOLD 3500  /* define high vs. low signal */
#define NUM_TEST_POINTS 4           /* number of test points in each bit */
#define NUM_CONTROL_BITS 11         /* number of control bits in each second*/
#define EXIT_BAD_BIT 64             /* exit code for a bad bit */
#define EXIT_MISPLACED_CTRL_BIT 65  /* exit code for misplaced control bit */
#define EXIT_UNEXPECTED_EOF 66      /* EOF in the middle of second of data */

/* keep using the same timeseries and decoded_bits arrays */
float timeseries[SAMPLE_RATE];
int decoded_bits[BITS_PER_SECOND];
int load1sec();
int readbit(int i);
void validate_timeseries();
void print_time();
char *get_time_string(int sec, int min, int hrs, int day, int yr);
int exit_status = 0;

/* count characters in input; 1st version */
int main() {
    while (load1sec()) {
        validate_timeseries();
        if (exit_status != 0)
            return exit_status;
        print_time();
    }
    return 0;
}

/*
make sure that all bits were read correctly and that control codes are
in the correct places.
*/
void validate_timeseries() {
    register int i;
    const static int control_bit_addresses[NUM_CONTROL_BITS] = 
        {0,9,19,29,39,49,59,69,79,89,99};
    /* a value of 1 indicates a control bit */
    static int control_bits[BITS_PER_SECOND]; /* static initializes to zero */
    for (i=0; i<NUM_CONTROL_BITS; i++)
        control_bits[control_bit_addresses[i]] = 1;

    for (i=0; i<BITS_PER_SECOND; i++)
        if (decoded_bits[i] == -1) {
            puts("ERROR: Bad Bit! Dumping decoded bit stream:\n");
            for (i=0; i<BITS_PER_SECOND; i++)
                printf("%d\t%d\n", i, decoded_bits[i]);
            exit_status = EXIT_BAD_BIT;
            break;
        } else if (control_bits[i] && decoded_bits[i] != 2) {
            exit_status = EXIT_MISPLACED_CTRL_BIT;
            puts("ERROR: Misplaced Control Bit at position ");
            printf("%d", i);
            puts("! Dumping decoded bit stream:\n");
            for (i=0; i<BITS_PER_SECOND; i++)
                printf("%d\t%d\t%d\n", i, control_bits[i], decoded_bits[i]);
            break;
        }
}

/* turn the decoded bits into a time and print that TODO make it formatted */
void print_time() {
    int sec, min, hrs, day, yr;
    sec = min = hrs = day = yr = 0;

    /* how many seconds does each bit represent? */
    sec += 1 * decoded_bits[1];
    sec += 2 * decoded_bits[2];
    sec += 4 * decoded_bits[3];
    sec += 8 * decoded_bits[4];
    sec += 10 * decoded_bits[6];
    sec += 20 * decoded_bits[7];
    sec += 40 * decoded_bits[8];

    /* how many minutes does each bit represent? */
    min += 1 * decoded_bits[10];
    min += 2 * decoded_bits[11];
    min += 4 * decoded_bits[12];
    min += 8 * decoded_bits[13];
    min += 10 * decoded_bits[15];
    min += 20 * decoded_bits[16];
    min += 40 * decoded_bits[17];

    /* how many hours does each bit represent? */
    hrs += 1 * decoded_bits[20];
    hrs += 2 * decoded_bits[21];
    hrs += 4 * decoded_bits[22];
    hrs += 8 * decoded_bits[23];
    hrs += 10 * decoded_bits[25];
    hrs += 20 * decoded_bits[26];

    /* how many days does each bit represent? */
    day += 1 * decoded_bits[30];
    day += 2 * decoded_bits[31];
    day += 4 * decoded_bits[32];
    day += 8 * decoded_bits[33];
    day += 10 * decoded_bits[35];
    day += 20 * decoded_bits[36];
    day += 40 * decoded_bits[37];
    day += 80 * decoded_bits[38];
    day += 100 * decoded_bits[40];
    day += 200 * decoded_bits[41];

    /* how many years does each bit represent? */
    yr += 1 * decoded_bits[50];
    yr += 2 * decoded_bits[51];
    yr += 4 * decoded_bits[52];
    yr += 8 * decoded_bits[53];
    yr += 10 * decoded_bits[55];
    yr += 20 * decoded_bits[56];
    yr += 40 * decoded_bits[57];
    yr += 80 * decoded_bits[58];

    /* print the current time */
    printf("%s", get_time_string(sec, min, hrs, day, yr));
}

/*
get a string representing time based on seconds, minutes, hours, days, and
years since 2000
*/
char *get_time_string(int sec, int min, int hrs, int day, int yr) {
    struct tm tstruct = {
        sec,        /* seconds */
        min,        /* minutes */
        hrs,        /* hours */
        day,        /* days since start of year, i.e. since start of january */
        0,          /* month of january */
        100 + yr,   /* years since 1900 */
        0,          /* days since start of week (not used) */
        0,          /* days since start of year (not used) */
        -1          /* daylight savings time, unknown */
    };
    /* properly calculate month and day for display */
    mktime(&tstruct);
    /* return date in format Sun Jan  3 15:14:13 1988\n */
    return asctime(&tstruct);
}

/* this will only process 1 second worth of data */
int load1sec() {
    int i, j;
    
    /*
    load the timeseries from stdin.
    return 0 if done reading.
    */
    for (i=0; i<SAMPLE_RATE; i++)
        if (fscanf(stdin, "%f\n", &timeseries[i]) == EOF) {
            if (i != 0)
                exit_status = EXIT_UNEXPECTED_EOF;
            return 0;
        }

    /*
    find beginning of each IRIG-B signal bit (they are evenly spaced)
    and decode that bit.
    */
    for (i=0; i<BITS_PER_SECOND; i++)
        decoded_bits[i] = readbit((i * SAMPLE_RATE) / BITS_PER_SECOND);

    return 1;
}

/*
Read each bit and return its value: 0, 1, -1 (error), or 2 (control bit).
How each character is represented at each test point:
BIT VALUE:          REPRESENTATION:
0                   [1,0,0,0]
1                   [1,1,0,0]
2  (control)        [1,1,1,0]
-1 (error)          (anything else)
*/
int readbit(int i) {
    /* how many seconds into each bit to measure? round it. */
    register int j;
    static int tp[NUM_TEST_POINTS];
    const static int TEST_POINTS[NUM_TEST_POINTS] = {
        (int) SAMPLE_RATE*1.22e-03 + 0.5,
        (int) SAMPLE_RATE*3.66e-03 + 0.5,
        (int) SAMPLE_RATE*6.71e-03 + 0.5,
        (int) SAMPLE_RATE*9.16e-03 + 0.5
    };

    /* get values at test points; will be 0 or 1 */
    for (j=0; j<NUM_TEST_POINTS; j++)
        tp[j] = timeseries[i+TEST_POINTS[j]] > HIGH_SIGNAL_THRESHOLD;
    /* TODO remove debug */
    /* printf("[%d %d %d %d]\n", tp[0], tp[1], tp[2], tp[3]); */

    /* test for invalid values */
    if (tp[1] < tp[2] || tp[0] == 0 || tp[3] == 1)
        return -1;

    /* decode */
    return tp[1] + tp[2];
}

File: geco_irig_decode.py
#!/usr/bin/env python
# (c) Stefan Countryman, 2016-2017

import sys
from datetime import datetime, timedelta
from textwrap import fill
import numpy as np
# import scipy.ndimage.filters as scf

if __name__ == "__main__" and len(sys.argv) > 1:
    print(
        "Usage: {} <input_file.txt\n\n".format(sys.argv[0]) +
        fill(
            """Read a raw IRIG-B signal from STDIN and print out decoded
            timestamps.  Data must be a newline-delimited list of floating
            point values representing the value of the IRIG-B signal at each
            point in time.  Sample rate must be 16,384 (2^14) Hz. The input
            file must contain an integer number of seconds worth of data, i.e.
            it must have (Sample Rate) x (N) values, where N is the number of
            seconds that must be decoded, and the data must start at the
            beginning of a second.
            """
        )
    )
    exit(1)

# ------------------------------------------------------------------------------
# CONSTANTS
# ------------------------------------------------------------------------------

# max and min of histogram, and number of bins
SAMPLE_RATE = 16384     # ADC sample rate

# --------------------------------------------------------------------------
# IRIG-B RELATED CONSTANTS
# --------------------------------------------------------------------------
BITS_PER_SECOND = 100   # bits per second in IRIG-B spec
CONVOLUTION_SIGMA = 1e-4
HIGH_SIGNAL_THRESHOLD = 3500

# find the high/low test points for each type of bit (0, 1, or control);
# measured as indices from start of each bit
# TODO measure in fractions of a second and convert with SAMPLE_RATE
TEST_POINTS = np.array([20, 60, 110, 150])

# find the start of each bit
BIT_STARTS = np.round(np.arange(0, 1, 1./BITS_PER_SECOND) * SAMPLE_RATE)
ALL_TEST_POINT_INDICES = (BIT_STARTS[:, np.newaxis] + TEST_POINTS).astype(int)

# representations of each type of bit (0, 1, or control)
REP_0 = [1, 0, 0, 0]
REP_1 = [1, 1, 0, 0]
REP_C = [1, 1, 1, 0]
CURRENT_CENTURY = 20

# how many seconds does each bit represent?
SECONDS = np.zeros(BITS_PER_SECOND, dtype=int)
SECONDS[1] = 1
SECONDS[2] = 2
SECONDS[3] = 4
SECONDS[4] = 8
SECONDS[6] = 10
SECONDS[7] = 20
SECONDS[8] = 40

# how many minutes does each bit represent?
MINUTES = np.zeros(BITS_PER_SECOND, dtype=int)
MINUTES[10] = 1
MINUTES[11] = 2
MINUTES[12] = 4
MINUTES[13] = 8
MINUTES[15] = 10
MINUTES[16] = 20
MINUTES[17] = 40

# how many hours does each bit represent?
HOURS = np.zeros(BITS_PER_SECOND, dtype=int)
HOURS[20] = 1
HOURS[21] = 2
HOURS[22] = 4
HOURS[23] = 8
HOURS[25] = 10
HOURS[26] = 20

# how many days does each bit represent?
DAYS = np.zeros(BITS_PER_SECOND, dtype=int)
DAYS[30] = 1
DAYS[31] = 2
DAYS[32] = 4
DAYS[33] = 8
DAYS[35] = 10
DAYS[36] = 20
DAYS[37] = 40
DAYS[38] = 80
DAYS[40] = 100
DAYS[41] = 200

# how many years does each bit represent?
YEARS = np.zeros(BITS_PER_SECOND, dtype=int)
YEARS[50] = 1
YEARS[51] = 2
YEARS[52] = 4
YEARS[53] = 8
YEARS[55] = 10
YEARS[56] = 20
YEARS[57] = 40
YEARS[58] = 80

control_bits = np.zeros(BITS_PER_SECOND, dtype=bool)
control_bits[range(9,100,10)] = True
control_bits[0] = True

#-------------------------------------------------------------------------------
# UTILITY FUNCTIONS
#-------------------------------------------------------------------------------

def decode_timeseries(timeseries):
    """Return the full decoded information as a dictionary along with a decoded
    datetime object for more convenient manipulation."""
    # filter the timeseries to remove ringing at corners
    # filt = scf.gaussian_filter1d(timeseries, CONVOLUTION_SIGMA * SAMPLE_RATE)

    # check all test points
    bits_high = (timeseries[ALL_TEST_POINT_INDICES] >= HIGH_SIGNAL_THRESHOLD).astype(int)

    # represent control bits with the number 2
    bits = np.zeros(BITS_PER_SECOND, dtype=int)
    for i in range(bits_high.shape[0]):
        if   np.all(bits_high[i] == REP_0): bits[i] = 0
        elif np.all(bits_high[i] == REP_1): bits[i] = 1
        elif np.all(bits_high[i] == REP_C): bits[i] = 2
        else: raise ValueError("Bad bit: " + str(i))

    # are the control bits in the correct spots?
    if not np.all((bits == 2) == control_bits):
        raise ValueError("Control bits are not present where expected: \n"
                        + str((bits == 2) == control_bits))

    # find total seconds, minutes, hours, days, and years
    decoded = {
        'second': bits.dot(SECONDS),
        'minute': bits.dot(MINUTES),
        'hour': bits.dot(HOURS),
        'day': bits.dot(DAYS),
        'year': bits.dot(YEARS) + 100*CURRENT_CENTURY,
    }

    # parse a datetime from this
    jan1 = datetime(decoded['year'], 1, 1, decoded['hour'], decoded['minute'],
                    decoded['second'])
    decoded['datetime'] = jan1 + timedelta(decoded['day'] - 1)

    return decoded

def get_date_from_timeseries(timeseries):
    """Decode the input waveform, which is assumed to be a 16384hz digitized
    IRIG-B signal using DCLS (DC Level Shift)."""
    return decode_timeseries(timeseries)['datetime']

def print_formatted_date(converted_date):
    # finally, print the date
    print(converted_date.strftime('%a %b %d %X %Y'))

# and will cause a ValueError.
def read_1_second_from_stdin():
    # read in data from stdin; don't read more than a second worth of data
    timeseries = np.zeros(SAMPLE_RATE)
    line = ''
    i = 0
    while i < SAMPLE_RATE:
        line = sys.stdin.readline()
        if not line:
            if i == 0:
                raise EOFError('Hit EOF at end of a second.')
            else:
                raise ValueError('Hit EOF ' + str(i) + ' lines into second; '
                                 'provide integer number of seconds of data.')
        timeseries[i] = float(line)
        i += 1
    return timeseries

def main():
    while True:
        try:
            timeseries = read_1_second_from_stdin()
            print_formatted_date(get_date_from_timeseries(timeseries))
        except EOFError:
            return

# run this if we are running from command line
if __name__ == "__main__":
    main()
